File: Difficulty: Hard/Maximum XOR subarray/maximum-xor-subarray.cs
//{ Driver Code Starts
//Initial Template for C#


using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DriverCode
{

    class GFG
    {
        static void Main(string[] args)
        {
            int testcases;// Taking testcase as input
            testcases = Convert.ToInt32(Console.ReadLine());
            while (testcases-- > 0)// Looping through all testcases
            {
                int N = Convert.ToInt32(Console.ReadLine());
                int[] arr = new int[N];

                string elements = Console.ReadLine().Trim();
                elements = elements + " ";
                int k = 0;
                string element = "";
                for (int j = 0; j < elements.Length; j++)
                {

                    if (elements[j] == ' ')
                    {
                        arr[k] = int.Parse(element);
                        element = "";
                        k++;
                    }
                    else
                    {
                        element = element + elements[j];
                    }
                }
                Solution obj = new Solution();
                int res = obj.maxSubarrayXOR( N, arr);
                Console.Write(res+"\n");
          
Console.Write("~"+"\n");
}

        }
    }
}



// } Driver Code Ends


//User function Template for C#

class Solution {

    // Make TrieNode public so it can be accessed by public methods
    public class TrieNode {
        public TrieNode[] child = new TrieNode[2];
    }

    // Function to insert a number into the Trie
    public void insert(TrieNode root, int num) {
        TrieNode node = root;
        for (int i = 31; i >= 0; i--) {
            int bit = (num >> i) & 1;
            if (node.child[bit] == null) {
                node.child[bit] = new TrieNode();
            }
            node = node.child[bit];
        }
    }

    // Function to find the maximum XOR of a number with elements in the Trie
    public int findMaxXOR(TrieNode root, int num) {
        TrieNode node = root;
        int maxXor = 0;
        
        for (int i = 31; i >= 0; i--) {
            int bit = (num >> i) & 1;
            int oppositeBit = bit ^ 1;
            
            if (node.child[oppositeBit] != null) {
                maxXor |= (1 << i);
                node = node.child[oppositeBit];
            } else {
                node = node.child[bit];
            }
        }
        
        return maxXor;
    }

    // Function to calculate the maximum subarray XOR
    public int maxSubarrayXOR(int N, int[] arr) {
        // Initialize root of Trie
        TrieNode root = new TrieNode();
        
        // Variable to keep track of the maximum XOR
        int maxXor = int.MinValue;
        
        // To keep track of the prefix XOR
        int prefixXor = 0;
        
        // Insert 0 initially to handle case when prefixXor itself is the maximum
        insert(root, 0);
        
        // Traverse through the array
        for (int i = 0; i < N; i++) {
            // Update the current prefix XOR
            prefixXor ^= arr[i];
            
            // Find the maximum XOR with the current prefix XOR
            maxXor = Math.Max(maxXor, findMaxXOR(root, prefixXor));
            
            // Insert the current prefix XOR into the Trie
            insert(root, prefixXor);
        }
        
        return maxXor;
    }
}

File: Difficulty: Hard/Maximum XOR subarray/maximum-xor-subarray.js
//{ Driver Code Starts
//Initial Template for javascript

"use strict";

process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;

process.stdin.on("data", (inputStdin) => {
  inputString += inputStdin;
});

process.stdin.on("end", (_) => {
  inputString = inputString
    .trim()
    .split("\n")
    .map((string) => {
      return string.trim();
    });

  main();
});

function readLine() {
  return inputString[currentLine++];
}

/* Function to print an array */
function printArray(arr, size) {
  let i;
  let s = "";
  for (i = 0; i < size; i++) {
    if(arr[i] == -0)
      arr[i] = 0;
    s += arr[i] + " ";
  }
  console.log(s);
}

function main() {
  let t = parseInt(readLine());
  let i = 0;
  for (; i < t; i++) {
    let n = parseInt(readLine());
    let arr = new Array(n);
    let input = readLine().split(" ").map((x)=>parseInt(x));
    for(let j = 0;j<n;j++) arr[j] = input[j];   
    let obj = new Solution();
    let res = obj.maxSubarrayXOR(arr,n);
    console.log(res);
  
console.log("~");
}
}
// } Driver Code Ends


//User function Template for javascript

/**
 * @param {number[]} arr
 * @param {number} n
 * @return {number}
 */

class Solution {
    maxSubarrayXOR(arr, n) {
        // Trie node structure
        class TrieNode {
            constructor() {
                this.child = [null, null];  // child[0] for bit 0, child[1] for bit 1
            }
        }

        // Function to insert the number into the Trie
        const insert = (root, num) => {
            let node = root;
            for (let i = 31; i >= 0; i--) {
                let bit = (num >> i) & 1;
                if (!node.child[bit]) {
                    node.child[bit] = new TrieNode();
                }
                node = node.child[bit];
            }
        };

        // Function to find the maximum XOR of num with elements in the Trie
        const findMaxXOR = (root, num) => {
            let node = root;
            let maxXor = 0;
            for (let i = 31; i >= 0; i--) {
                let bit = (num >> i) & 1;
                let oppositeBit = bit ^ 1;
                if (node.child[oppositeBit]) {
                    maxXor |= (1 << i);
                    node = node.child[oppositeBit];
                } else {
                    node = node.child[bit];
                }
            }
            return maxXor;
        };

        // Initialize Trie root and other variables
        let root = new TrieNode();
        insert(root, 0);  // Insert 0 initially to handle the case when prefixXor itself is the maximum
        let maxXor = -Infinity;
        let prefixXor = 0;

        // Traverse through the array
        for (let i = 0; i < n; i++) {
            // Update the current prefix XOR
            prefixXor ^= arr[i];
            
            // Find the maximum XOR with the current prefix XOR
            maxXor = Math.max(maxXor, findMaxXOR(root, prefixXor));
            
            // Insert the current prefix XOR into the Trie
            insert(root, prefixXor);
        }
        
        return maxXor;
    }
}
